src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp

```diff
@@ -95,7 +904,1269 @@ class boss_thorim : public CreatureScript
+
+                bool operator()(Unit* u)
+                {
+                    if (_exclSelf && u == _referer)
+                        return false;
+
+                    if (_exclAura && u->HasAura(sSpellMgr->GetSpellIdForDifficulty(_exclAura, _referer)))
+                        return false;
+
+                    if ((u->GetHealth() + GetRemainingHealOn(u) + _hp) > u->GetMaxHealth())
+                        return false;
+
+                    uint32 missingHP = u->GetMaxHealth() - u->GetHealth();
+                    if (u->IsAlive() && _referer->IsFriendlyTo(u) && _referer->IsWithinDistInMap(u, _range) && missingHP > _hp)
+                    {
+                        _hp = missingHP;
+                        return true;
+                    }
+
+                    return false;
+                }
+
+            private:
+                Unit const* _referer;
+                float _range;
+                uint32 _hp;
+                uint32 _exclAura;
+                bool _exclSelf;
+        };
+
+        static Unit* GetUnitWithMostMissingHp(SpellInfo const* spellInfo, Unit* caster)
+        {
+            float range = spellInfo->GetMaxRange(false);
+            uint32 heal = GetTotalHeal(spellInfo, caster);
+
+            Unit* target = nullptr;
+            MostHPMissingInRange checker(caster, range, heal);
+            Trinity::UnitLastSearcher<MostHPMissingInRange> searcher(caster, target, checker);
+            caster->VisitNearbyObject(range, searcher);
+
+            return target;
+        }
+
+        static Unit* GetHealTarget(uint32 spellId, Unit* caster)
+        {
+            Unit* healTarget = nullptr;
+
+            if (SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(spellId))
+            {
+                if (!spellInfo->HasAttribute(SPELL_ATTR1_CANT_TARGET_SELF) && !roll_chance_f(caster->GetHealthPct()) && !((caster->GetHealth() + GetRemainingHealOn(caster) + GetTotalHeal(spellInfo, caster)) > caster->GetMaxHealth()))
+                    healTarget = caster;
+                else
+                    healTarget = GetUnitWithMostMissingHp(spellInfo, caster);
+            }
+
+            return healTarget;
+        }
+    };
+
+    bool UseAbility(uint32 spellId)
+    {
+        Unit* target = nullptr;
+        if (AIHelper::GetTotalHeal(spellId, me))
+            target = AIHelper::GetHealTarget(spellId, me);
+        else
+            target = me->GetVictim();
+
+        if (target)
+        {
+            if (_info->Type == MERCENARY_SOLDIER)
+            {
+                bool allowMove = true;
+                if (SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(spellId))
+                {
+                    if (me->IsInRange(target, spellInfo->GetMinRange(false), spellInfo->GetMaxRange(false))
+                        && me->IsWithinLOSInMap(target))
+                        allowMove = false;
+                }
+                SetCombatMovement(allowMove);
```

Comment:
correction, will work since there is currently an infinite loop in UpdateVictim() (on each update tick a new movement is created), so better do this correctly since that will be fixed (soon 💃 )